discovery_engine_core/web-api/src/bin/ingestion.rs

```diff
@@ -170,6 +170,18 @@ fn init_model(config: &Config) -> Result<Model, GenericError> {
     Ok(Arc::new(smbert))
 }
 
+macro_rules! handle4 {
+    ($handle:ident) => {
+        |t1, t2, t3, t4| async move {
+            let (span, id) = setup_request_span(stringify!($handle));
+            $handle(t1, t2, t3, t4)
+                .instrument(span)
+                .await
+                .or_else(|err| handle_rejection(err, id))
+        }
+    };
+}
```

Comment:
This hackery is necessary, it also means that depending on the number of parameters we will also need handle1,2,3,4,5 etc.

The reason this is necessary is due to serve!! limitations of warp:

1. `warp::trace::*` can only access a small number of fixed information, not including any parameters, extensions and similar
2. `filter.recover` can only depend on the rejection, there is no way to inject the log id here
3. due to non-exposed pub traits in warp you can not use the abstraction they use to allow functions with various parameter, elimination of () etc.
4. due to limitations of the rust type inference you can't make this a function instead of a macro
5. due to limitations of warp you can't create wrappers or filter from scratch, you can only create them through existing combinators which is not use-full if the existing combinators don't provide what you need
6. once you have a "Rejection" error result (e.g. because of an extractor of a parameter) you can't go back to having a custom error due to limitations in warp
7. you can't properly access hyper;s extension system nor can you include tower layers into warp due to limitations in it's design
8. the return reply of all endpoint handles and `handle_rejections` must be `Box<dyn Reply>` as due to limitations of warp you can't have an either type (and the either type warp has internally is not exposed). Also you can't do nicely the boxing with a `map` on the handle/handle_reject calls due to rust limitations of impl trait.
9. There was an initial bug where any error from middle ware (e.g. wrong content type) got turned into internal server error.
10. :pushpin: Now this is fixed but there is an error that such errors return a string instead of json body
11. It is not trivial to fix this due to what `Rejection` doesn't expose to consumers of the library
11. :pushpin: there is a bug where additional path segments might be accepted due to missing `path::end()` which is really easy to forget
12. While we now can have the `log_id` we in turn can no longer access the path, http method or http version, I used the handlers function name as a work around.

Reply:
This seems horrible enough to make me reconsider the use of warp. Are we missing something?

Reply:
i might have found a solution to this problem, pls have a look if it fits your needs for the follow up prs:
- the instrumentation can be done via the attribute macro, such that we don't have to manage the span manually
- the `handle_rejection()` can be combined via an `or_else()` with the previous filter, such that there is no need for argument-size-dependent macros (the success result type has to be wrapped in a 1-tuple to satisfy the trait bounds)
- the `log_id` is then recorded as a field to the request handler and passed to the rejection handler, overall it looks like:

```rust
// POST /add_data
#[instrument(
    skip_all,
    level = "info",
    name = "request",
    fields(method = "handle_add_data", log_id)
)]
fn post_add_data(
    config: Config,
    model: Model,
    client: Client,
) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone {
    let log_id = Uuid::new_v4();
    Span::current().record("log_id", log_id.to_string());
    debug!("received request");

    warp::path("add_data")
        .and(warp::post())
        .and(warp::body::content_length_limit(config.max_body_size))
        .and(warp::body::json())
        .and(with_model(model))
        .and(with_config(config))
        .and(with_client(client))
        .and_then(handle_add_data)
        .or_else(move |err| handle_rejection(err, log_id))
}
```
where
```rust
async fn handle_rejection(err: Rejection, log_id: Uuid) -> Result<(Box<dyn Reply>,), Rejection> {
    // ... body ...
    Ok((Box::new(warp::reply::with_status(json, code)),))
}
```

wdyt, @rustonaut & @acrrd ?

Reply:
This will add a bit of boilerplate in each endpoint but is nicer then the macro.

Reply:
this does not work, we need a new uuid/spawn for each request this creates a single uuid/spawn when creating route mapping.

Reply:
Like or else would need to  accept a `|err, uuid| {}` function. This it doesn't. So I thought to make it accept an `|(err, uuid)|` and make this the error type, but you can't go from having a `Rejection` to a custom error. There is no `map_err` and `or_else` is restricted to returning `IsReject` which can't be implemented by custom types it only allows `Infallible` instead of `Rejection` (it's sealed, also I think `pub` but not `pub` exported).

Reply:
There is also the problem error handling via `recover` or `or_else` requires you to know every specific error type which an rejection might be created from to allow an reliable consistent error return value.

Not hard for our code as we can just have a custom error we defined and use that and only turn that into `Rejection`, but a problem because we also have to know that about every error type of every middle ware. I.e. without the `handle` macro we turn "wrong content type" or even 404 errors into internal server errors :face_exhaling:

Reply:
hAlso in case we continue to use warp:

`warp::path("add_data")` needs to be `warp::path("add_data").and(warp::path::end())` or `warp::path!("add_data")`.

Here `warp::path()` is a hidden reexport of `warp::path::path()` and not quite the same as `warp::path!()` which auto includes an `end` but doesn't allow combining it with known path parts (as we use it for user).

Not including `end` means it will match `/add_data` and any `/add_data/*`.